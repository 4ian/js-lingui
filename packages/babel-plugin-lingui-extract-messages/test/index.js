import fs from 'fs'
import path from 'path'
import { transformFileSync } from 'babel-core'

import plugin from '../src/index'

const LOCALE_DIR = './locale'

const rmdir = (dir) => {
  if (!fs.existsSync(dir)) return
  const list = fs.readdirSync(dir)

  for (let i = 0; i < list.length; i++) {
    const filename = path.join(dir, list[i])
    const stat = fs.statSync(filename)

    if (filename === '.' || filename === '..') {
      // pass these files
    } else if (stat.isDirectory()) {
      // rmdir recursively
      rmdir(filename)
    } else {
      // rm fiilename
      fs.unlinkSync(filename)
    }
  }
  fs.rmdirSync(dir)
}

function testCase (testName, assertion) {
  const transform = (filename, jsx = true) => () => transformFileSync(path.join(__dirname, 'fixtures', filename), {
    babelrc: false,
    plugins: [
      ...(filename.endsWith('integration.js')
          ? jsx
            ? [
              'lingui-transform-react',
              'lingui-transform-js'
            ]
            : ['lingui-transform-js']
          : []
      ),
      [plugin, {
        localeDir: LOCALE_DIR
      }],
      'external-helpers',
      ...(jsx ? ['syntax-jsx'] : []),
      'transform-remove-strict-mode'
    ]
  })

  it(testName, () => assertion(transform))
}

describe('babel-plugin-lingui-extract-messages', function () {
  // CWD is root directory of repository, so origin of all messages is going to
  // relative to root
  const buildDir = path.join(
    LOCALE_DIR,
    '_build', 'packages', 'babel-plugin-lingui-extract-messages',
    'test', 'fixtures'
  )

  beforeAll(() => {
    rmdir(LOCALE_DIR)
  })

  afterAll(() => {
    rmdir(LOCALE_DIR)
  })

  testCase('should raise exception on duplicate id and different defaults', (transform) => {
    expect(transform('jsx/duplicate-id.js')).toThrow(/Different defaults/)
  })

  testCase("shouldn't write catalog for files without translatable messages", (transform) => {
    expect(transform('empty.js')).not.toThrow()
    expect(fs.existsSync(path.join(buildDir, 'empty.json'))).toBeFalsy()
  })

  testCase("shouldn't path to file inside locale dir", (transform) => {
    expect(transform('jsx/deep/all.js')).not.toThrow()
    expect(fs.existsSync(path.join(buildDir, 'jsx/deep/all.json'))).toBeTruthy()
  })

  testCase('should extract all messages from JSX files', (transform) => {
    // first run should create all required folders and write messages
    expect(transform('jsx/all.js')).not.toThrow()
    // another runs should just write messages
    expect(transform('jsx/all.js')).not.toThrow()

    const messages = JSON.parse(fs.readFileSync(path.join(buildDir, 'jsx/all.json')))
    expect(messages).toMatchSnapshot()
  })

  testCase('should extract all messages from JSX files (integration)', (transform) => {
    // first run should create all required folders and write messages
    expect(transform('jsx/integration.js')).not.toThrow()
    // another runs should just write messages
    expect(transform('jsx/integration.js')).not.toThrow()

    const messages = JSON.parse(fs.readFileSync(path.join(buildDir, 'jsx/integration.json')))
    expect(messages).toMatchSnapshot()
  })

  testCase('should extract all messages from JS files', (transform) => {
    // first run should create all required folders and write messages
    expect(transform('js/all.js', false)).not.toThrow()
    // another runs should just write messages
    expect(transform('js/all.js', false)).not.toThrow()

    const messages = JSON.parse(fs.readFileSync(path.join(buildDir, 'js/all.json')))
    expect(messages).toMatchSnapshot()
  })

  testCase('should extract all messages from JS files (integration)', (transform) => {
    // first run should create all required folders and write messages
    expect(transform('js/integration.js', false)).not.toThrow()
    // another runs should just write messages
    expect(transform('js/integration.js', false)).not.toThrow()

    const messages = JSON.parse(fs.readFileSync(path.join(buildDir, 'js/integration.json')))
    expect(messages).toMatchSnapshot()
  })
})
