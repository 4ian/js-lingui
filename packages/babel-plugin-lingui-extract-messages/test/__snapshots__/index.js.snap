exports[`babel-plugin-lingui-extract-messages should extract all messages from JS files (integration) 1`] = `
Object {
  "msg.hello": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/js/integration.js",
        null,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JS files 1`] = `
Object {
  "msg.default": Object {
    "defaults": "Hello World",
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/js/all.js",
        5,
      ],
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/js/all.js",
        6,
      ],
    ],
  },
  "msg.hello": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/js/all.js",
        4,
      ],
    ],
  },
  "{count, plural, one {# book} other {# books}}": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/js/all.js",
        7,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files (integration) 1`] = `
Object {
  "Hello World": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/integration.js",
        2,
      ],
    ],
  },
  "Hi, my name is {name}": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/integration.js",
        3,
      ],
    ],
  },
  "Title": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/integration.js",
        null,
      ],
    ],
  },
  "{undefined, plural, count {undefined} one {# book} others {# books}}": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/integration.js",
        null,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files 1`] = `
Object {
  "Hi, my name is <0>{name}</0>": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/all.js",
        5,
      ],
    ],
  },
  "msg.default": Object {
    "defaults": "Hello World",
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/all.js",
        3,
      ],
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/all.js",
        4,
      ],
    ],
  },
  "msg.hello": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/all.js",
        2,
      ],
    ],
  },
  "{count, plural, one {# book} other {# books}}": Object {
    "origin": Array [
      Array [
        "packages/babel-plugin-lingui-extract-messages/test/fixtures/jsx/all.js",
        6,
      ],
    ],
  },
}
`;
